dask/dataframe/partitionquantiles.py

```diff
@@ -0,0 +1,456 @@
+    # 3. Concatenate the values from (1) and (2), sort, and return.
+    #
+    # We assume that all values are unique, which happens in the previous
+    # step `merge_and_compress_summaries`.
+
+    if len(vals) == npartitions + 1:
+        rv = vals
+    elif len(vals) < npartitions + 1:
+        # The data is under-sampled
+        if np.issubdtype(vals.dtype, np.number):
+            # Interpolate extra divisions
+            q_weights = np.cumsum(weights)
+            q_target = np.linspace(q_weights[0], q_weights[-1], npartitions + 1)
+            rv = np.interp(q_target, q_weights, vals)
+        else:
+            # Distribute the empty partitions
+            duplicated_index = np.linspace(
+                0, len(vals) - 1, npartitions - len(vals) + 1, dtype=int
+            )
+            duplicated_vals = vals[duplicated_index]
+            rv = np.concatenate([vals, duplicated_vals])
+            rv.sort()
+    else:
+        target_weight = weights.sum() / npartitions
+        jumbo_mask = weights >= target_weight
+        jumbo_vals = vals[jumbo_mask]
+
+        trimmed_vals = vals[~jumbo_mask]
+        trimmed_weights = weights[~jumbo_mask]
+        trimmed_npartitions = npartitions - len(jumbo_vals)
+
+        # percentile-like, but scaled by weights
+        q_weights = np.cumsum(trimmed_weights)
+        q_target = np.linspace(0, q_weights[-1], trimmed_npartitions + 1)
+
+        left = np.searchsorted(q_weights, q_target, side='left')
+        right = np.searchsorted(q_weights, q_target, side='right') - 1
+        # stay inbounds
+        np.maximum(right, 0, right)
+        lower = np.minimum(left, right)
+        trimmed = trimmed_vals[lower]
+
+        rv = np.concatenate([trimmed, jumbo_vals])
+        rv.sort()
+
+    if str(dtype) == 'category':
+        rv = pd.Categorical.from_codes(rv, info[0], info[1])
+    elif rv.dtype != dtype:
+        rv = rv.astype(dtype)
+    return rv
+
+
+def percentiles_summary(df, num_old, num_new, upsample=1.0, random_state=None):
+    """Summarize data using percentiles and derived weights.
+
+    These summaries can be merged, compressed, and converted back into
+    approximate percentiles.
+
+    Parameters
+    ----------
+    df: pandas.Series
+        Data to summarize
+    num_old: int
+        Number of partitions of the current object
+    num_new: int
+        Number of partitions of the new object
+    upsample: float
+        Scale factor to increase the number of percentiles calculated in
+        each partition.  Use to improve accuracy.
+    """
+    from dask.array.percentile import _percentile
+    length = len(df)
+    qs = sample_percentiles(num_old, num_new, length, upsample, random_state)
+    data = df.values
+    interpolation = 'linear'
+    if str(data.dtype) == 'category':
+        data = data.codes
+        interpolation = 'nearest'
+    vals = _percentile(data, qs, interpolation=interpolation)
```

Comment:
A lingering thought: should we always use `interpolation='nearest'`, which may behave more nicely with our simple compression step that combines equal values?  Doing linear interpolation pretty much makes our compression step useless.